### PR TITLE
Fix: [CDS-67964]: Matrix nodes get copied while retrying different st…

### PR DESCRIPTION
**Pull request by yagyanshbhatia**
## Describe your changes

**Changes in NextStepHandler**: This change would change the output for createIdentityNodeIfRequired. This is required because if we use the same node, all such nodes will point to the same identity plan node. Overriding changes to code flow to allow creating a different identity plan node for all executions. 

**Changes in IdentityStrategyStep**: Current node needs to be added into children as execution node only if its part of the retry stage. If we see a failed step that isn't part of the retry stage, it should be added as an identity stage. This allows us to create an identity node for all such executions and not just use the same IdentityPlanNode pointing to one of the executions (hence copying the status). 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47857)

### CODE REVIEW

File: pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java [hunk 1]
```diff
@@ -91,23 +92,24 @@ private List<ChildrenExecutableResponse.Child> getChildrenFromNodeExecutions(
     List<ChildrenExecutableResponse.Child> children = new ArrayList<>();
     List<Node> identityNodesToBeCreated = new ArrayList<>();
     for (NodeExecution nodeExecution : childrenNodeExecutions) {
-      if (StatusUtils.brokeAndAbortedStatuses().contains(nodeExecution.getStatus())) {
+      Node node = planService.fetchNode(nodeExecution.getPlanId(), nodeExecution.getNodeId());
```
Comment (naidu-harness) [inline pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java:95]: can we add the PR description here as a comment

**Comment by brijeshdhakar** [comment]
trigger uts

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47857)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java` line 99 at r1](https://reviewable.io/reviews/harness/harness-core/47857#-NVbxRVWEd6pXx9D2LuA:-NVbxRVWEd6pXx9D2LuB:b-v758r5) ([raw file](https://github.com/harness/harness-core/blob/dfa2089c671696372d33289f641cf355ae445aff/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java#L99)):*
> ```Java
>   }
> 
>   boolean checkIfSameNodeIsRequired(Node nextNode, ExecutionMode executionMode) {
> ```

In this one , we needed to test without matrix also its working as expected or not, thus we should comments as discussed yesterday
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java` line 95 at r1](https://reviewable.io/reviews/harness/harness-core/47857#gh-1194602069:-NVbxLhH3OhGpfVH5PWo:binx7vo) ([raw file](https://github.com/harness/harness-core/blob/dfa2089c671696372d33289f641cf355ae445aff/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java#L95)):*
<details><summary><i>Previously, naidu-harness (Sanyasi Naidu Annepu) wrote…</i></summary><blockquote>

can we add the PR description here as a comment
</blockquote></details>

@yagyanshbhatia as discussed lets add comments

File: pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java (hunk 1 above)
Comment (yagyanshbhatia) [reply pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java:99]: Done.

**Comment by yagyanshbhatia** [comment]
trigger uts

**Comment by yagyanshbhatia** [comment]
> 

Done

**Comment by yagyanshbhatia** [comment]
> > Reviewed 1 of 2 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/47857)_ status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained.
> 
> _[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java` line 99 at r1](https://reviewable.io/reviews/harness/harness-core/47857#-NVbxRVWEd6pXx9D2LuA:-NVbxRVWEd6pXx9D2LuB:b-v758r5) ([raw file](https://github.com/harness/harness-core/blob/dfa2089c671696372d33289f641cf355ae445aff/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java#L99)):_
> 
> > ```java
> >   }
> > 
> >   boolean checkIfSameNodeIsRequired(Node nextNode, ExecutionMode executionMode) {
> > ```
> 
> In this one , we needed to test without matrix also its working as expected or not, thus we should comments as discussed yesterday
> 
> _[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java` line 95 at r1](https://reviewable.io/reviews/harness/harness-core/47857#gh-1194602069:-NVbxLhH3OhGpfVH5PWo:binx7vo) ([raw file](https://github.com/harness/harness-core/blob/dfa2089c671696372d33289f641cf355ae445aff/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/identity/IdentityStrategyStep.java#L95)):_
> 
> Previously, naidu-harness (Sanyasi Naidu Annepu) wrote…
> @yagyanshbhatia as discussed lets add comments

Done

**Comment by yagyanshbhatia** [comment]
trigger uts

**Comment by yagyanshbhatia** [comment]
trigger checkstylecodeformat

**Comment by yagyanshbhatia** [comment]
trigger uts

**Review by namanvermah — APPROVED** [review]

**Comment by yagyanshbhatia** [comment]
trigger uts

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47857)* status: 1. unresolved discussions. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java` line 99 at r1](https://reviewable.io/reviews/harness/harness-core/47857#-NVbxRVWEd6pXx9D2LuA:-NViBT3VBVvx6pFK2ym4:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/dfa2089c671696372d33289f641cf355ae445aff/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/advise/handlers/NextStepHandler.java#L99)):*
<details><summary><i>Previously, yagyanshbhatia (Yagyansh Bhatia) wrote…</i></summary><blockquote>

Done
</blockquote></details>

done

**Comment by yagyanshbhatia** [comment]
trigger review

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47857)* status: 1. unresolved discussions.

**Pull request merged**